src/layout/layout.rs

```diff
@@ -426,34 +437,83 @@ impl Layout {
 
         // create an element for each constraint that needs to be applied. Each element defines the
         // variables that will be used to compute the layout.
-        let elements = layout
+        let elements: Vec<Element> = layout
             .constraints
             .iter()
-            .map(|_| Element::new())
-            .collect::<Vec<Element>>();
-
-        // ensure that all the elements are inside the area
-        for element in &elements {
-            solver.add_constraints(&[
-                element.start | GE(REQUIRED) | area_start,
-                element.end | LE(REQUIRED) | area_end,
-                element.start | LE(REQUIRED) | element.end,
-            ])?;
-        }
+            .map(|_| Element::constrain(&mut solver, (area_start, area_end)))
+            .try_collect()?;
+
         // ensure there are no gaps between the elements
         for pair in elements.windows(2) {
             solver.add_constraint(pair[0].end | EQ(REQUIRED) | pair[1].start)?;
         }
-        // ensure the first element touches the left/top edge of the area
-        if let Some(first) = elements.first() {
-            solver.add_constraint(first.start | EQ(REQUIRED) | area_start)?;
-        }
-        if layout.segment_size != SegmentSize::None {
-            // ensure the last element touches the right/bottom edge of the area
-            if let Some(last) = elements.last() {
-                solver.add_constraint(last.end | EQ(REQUIRED) | area_end)?;
+
+        match layout.flex {
+            Flex::StretchLast => {
+                if let Some(first) = elements.first() {
```

Comment:
I wonder if there's a refactoring of this method that makes there always be an element (i.e. have one more element than constraints). This would make first and last always have a value, and remove all the if statements.

Reply:
This can and probably should be refactored, but I'd like to do it when I add more tests to document the implementation.